lib/cli/src/commands/run/mod.rs

```diff
@@ -128,7 +130,160 @@ impl Run {
+                            );
+                        }
+
+                        // Get backends that support these features
+                        tracing::info!("Filtering backends by feature support");
+                        let supported_backends =
+                            crate::backend::BackendType::filter_by_features(&detected_features);
+                        tracing::info!("Found {} compatible backends", supported_backends.len());
+
+                        if !supported_backends.is_empty() {
+                            // Find the highest priority backend that supports the detected features
+                            let mut feature_based_engine = None;
+
+                            // Try each backend
+                            for backend in &supported_backends {
+                                tracing::info!(
+                                    "Selecting backend {} which supports the required features",
+                                    backend
+                                );
+
+                                // Create engine with this specific backend
+                                let target = wasmer_compiler::types::target::Target::default();
+                                if let Ok(created_engine) =
+                                    backend.get_engine(&target, &detected_features)
+                                {
+                                    // We found a compatible engine
+                                    feature_based_engine = Some(created_engine);
+                                    break;
+                                }
+                            }
+
+                            // If we found a feature-based engine, use it
+                            if let Some(feature_engine) = feature_based_engine {
+                                tracing::info!("Using feature-based engine selection");
+                                feature_engine
+                            } else {
+                                // No backend could create an engine, fall back to default
+                                self.rt.get_engine()?
+                            }
+                        } else {
+                            // No backend supports all required features, use default
+                            self.rt.get_engine()?
+                        }
+                    }
+                }
+
+                #[cfg(not(feature = "compiler"))]
+                {
+                    // Without compiler feature, just use default engine
+                    self.rt.get_engine()?
+                }
+            }
+            None => {
+                // No WebAssembly file available for analysis, use default engine selection
+                self.rt.get_engine()?
+            }
+        };
+
         let be_kind = engine.get_backend_kind();
         tracing::info!("Executing on backend {be_kind:?}");
```

Comment:
Since it's not entirely obvious which backend actually runs a module, I'd suggest to change this to 
```suggestion
        tracing::info!("Executing on backend {}", engine.deterministic_id());
```
Otherwise llvm, singlepass and cranelift will be squashed under the same `SYS` backend.